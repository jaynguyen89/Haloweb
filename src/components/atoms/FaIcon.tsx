import { IconName } from '@fortawesome/free-regular-svg-icons';
import { IconDefinition } from '@fortawesome/free-solid-svg-icons';
import { faQuestion } from '@fortawesome/free-solid-svg-icons/faQuestion';
import { FontAwesomeIcon, FontAwesomeIconProps } from '@fortawesome/react-fontawesome';
import { useTheme } from '@mui/material';
import React, { useMemo } from 'react';

export type TIcon = Omit<FontAwesomeIconProps, 'icon'> & {
    wrapper: 'i' | 'fa',
    t?: 'str' | 'obj', // required when wrapper === 'fa'
    ic?: IconName | IconDefinition | string,
    variant?: 'fas' | 'far' | 'fal' | 'fab' | 'fat' | 'fad',
    animation?: 'beat' | 'beat-fade' | 'bounce' | 'fade' | 'flip' | 'shake' | 'spin' | 'spin-reverse' | 'spin-pulse', // only use when wrapper === 'i'
};

const FaIcon = ({
    wrapper,
    t,
    ic,
    variant = 'fas',
    animation,
    size = '1x',
    color,
    ...props
}: TIcon) => {
    const iconColor = color ?? useTheme().palette.primary.contrastText;

    if (wrapper === 'fa') {
        if (t === undefined || ic === undefined) return <FontAwesomeIcon icon={faQuestion} size='sm' color={iconColor} />;
        if (t === 'obj') return <FontAwesomeIcon {...props} size={size} icon={ic as IconDefinition} color={iconColor} />;

        return (
            <FontAwesomeIcon
                {...props}
                size={size}
                icon={[variant, ic as IconName]}
                color={iconColor}
            />
        );
    }

    const variantName = useMemo(() =>
        variant === 'fas' ? 'fa-solid'
        : variant === 'far' ? 'fa-regular'
        : variant === 'fal' ? 'fa-light'
        : variant === 'fab' ? 'fa-brands'
        : variant === 'fat' ? 'fa-thin'
        : 'fa-duotone',
        [variant],
    );

    const animationName = useMemo(() =>
        animation === undefined ? '' : `fa-${animation}`,
        [animation],
    );

    const styles = useMemo(() =>
        animation ? {color: iconColor, display: 'block'} : {color: iconColor},
        [animation],
    );

    return (
        <i className={`${variantName} fa-${ic} ${animationName} fa-${size}`} style={styles} />
    );
};

export default FaIcon;
