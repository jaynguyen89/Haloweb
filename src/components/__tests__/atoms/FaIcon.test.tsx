import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from '@cfaester/enzyme-adapter-react-18';
import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import toJson from 'enzyme-to-json';
import FaIcon, { TIcon } from 'src/components/atoms/FaIcon';
import { faTwitter } from '@fortawesome/free-brands-svg-icons/faTwitter';
import { faStar } from '@fortawesome/free-regular-svg-icons/faStar';

configure({ adapter: new Adapter() });

describe('FaIcon.tsx', () => {
    it('renders icon with `i` wrapper', () => {
        const propss: Array<TIcon> = [
            {
                wrapper: 'i',
                ic: 'pen',
                variant: 'fas',
                animation: 'beat',
                color: 'red',
                size: '2x',
            },
            {
                wrapper: 'i',
                ic: 'twitter',
                variant: 'fab',
                animation: 'fade',
                color: 'blue',
                size: '3x',
            },
            {
                wrapper: 'i',
                ic: 'comment',
                variant: 'far',
                animation: 'flip',
                color: 'green',
                size: '1x',
            },
            {
                wrapper: 'i',
                ic: 'star',
                variant: 'far',
                animation: 'spin',
                color: 'yellow',
                size: '2x',
            },
            {
                wrapper: 'i',
                ic: 'heart',
                variant: 'fas',
                animation: 'beat',
                color: 'orange',
                size: '5x',
            },
        ];

        const wrappers = propss.map(props => render(<FaIcon {...props} />));

        wrappers.forEach((wrapper, i) => {
            const tag = wrapper.container.getElementsByTagName('i');
            expect(tag.length).toBe(1);

            expect(tag[0].classList.value.includes(`fa-${propss[i].ic}`)).toBeTruthy();
            expect(tag[0].classList.value.includes(`fa-${propss[i].animation}`)).toBeTruthy();
            expect(tag[0].classList.value.includes(`fa-${propss[i].size}`)).toBeTruthy();

            const variant =
                propss[i].variant === 'fas' ? 'fa-solid' :
                propss[i].variant === 'fab' ? 'fa-brands' : 'fa-regular';
            expect(tag[0].classList.value.includes(variant)).toBeTruthy();

            expect(tag[0].style.color).toEqual(propss[i].color);
        });
    });

    it('renders icon with `fa` wrapper', () => {
        const propss: Array<TIcon> = [
            {
                wrapper: 'fa',
                t: 'str',
                ic: 'pen',
                variant: 'fas',
                animation: 'beat',
                color: 'red',
                size: '2x',
            },
            {
                wrapper: 'fa',
                t: 'obj',
                ic: faTwitter,
                animation: 'fade',
                color: 'blue',
                size: '3x',
            },
            {
                wrapper: 'fa',
                t: 'str',
                ic: 'comment',
                variant: 'far',
                animation: 'flip',
                color: 'green',
                size: '1x',
            },
            {
                wrapper: 'fa',
                t: 'obj',
                ic: faStar,
                animation: 'spin',
                color: 'yellow',
                size: '2x',
            },
            {
                wrapper: 'fa',
                ic: 'heart',
                variant: 'fas',
                animation: 'beat',
                color: 'orange',
                size: 'sm',
            },
        ];

        const wrappers = propss.map(props => shallow(<FaIcon {...props} />));
        wrappers.forEach(wrapper => expect(toJson(wrapper)).toMatchSnapshot());
    });
});
