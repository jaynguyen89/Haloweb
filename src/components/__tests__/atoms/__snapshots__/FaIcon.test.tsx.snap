// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FaIcon renders icon with \`fa\` wrapper 1`] = `
<FontAwesomeIcon
  beat={false}
  beatFade={false}
  border={false}
  bounce={false}
  className=""
  color="red"
  fade={false}
  fixedWidth={false}
  flip={false}
  icon={
    Array [
      "fas",
      "pen",
    ]
  }
  inverse={false}
  listItem={false}
  mask={null}
  maskId={null}
  pull={null}
  pulse={false}
  rotation={null}
  shake={false}
  size="2x"
  spin={false}
  spinPulse={false}
  spinReverse={false}
  swapOpacity={false}
  symbol={false}
  title=""
  titleId={null}
  transform={null}
/>
`;

exports[`FaIcon renders icon with \`fa\` wrapper 2`] = `
<FontAwesomeIcon
  beat={false}
  beatFade={false}
  border={false}
  bounce={false}
  className=""
  color="blue"
  fade={false}
  fixedWidth={false}
  flip={false}
  icon={
    Object {
      "icon": Array [
        512,
        512,
        Array [],
        "f099",
        "M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z",
      ],
      "iconName": "twitter",
      "prefix": "fab",
    }
  }
  inverse={false}
  listItem={false}
  mask={null}
  maskId={null}
  pull={null}
  pulse={false}
  rotation={null}
  shake={false}
  size="3x"
  spin={false}
  spinPulse={false}
  spinReverse={false}
  swapOpacity={false}
  symbol={false}
  title=""
  titleId={null}
  transform={null}
/>
`;

exports[`FaIcon renders icon with \`fa\` wrapper 3`] = `
<FontAwesomeIcon
  beat={false}
  beatFade={false}
  border={false}
  bounce={false}
  className=""
  color="green"
  fade={false}
  fixedWidth={false}
  flip={false}
  icon={
    Array [
      "far",
      "comment",
    ]
  }
  inverse={false}
  listItem={false}
  mask={null}
  maskId={null}
  pull={null}
  pulse={false}
  rotation={null}
  shake={false}
  size="1x"
  spin={false}
  spinPulse={false}
  spinReverse={false}
  swapOpacity={false}
  symbol={false}
  title=""
  titleId={null}
  transform={null}
/>
`;

exports[`FaIcon renders icon with \`fa\` wrapper 4`] = `
<FontAwesomeIcon
  beat={false}
  beatFade={false}
  border={false}
  bounce={false}
  className=""
  color="yellow"
  fade={false}
  fixedWidth={false}
  flip={false}
  icon={
    Object {
      "icon": Array [
        576,
        512,
        Array [
          11088,
          61446,
        ],
        "f005",
        "M287.9 0c9.2 0 17.6 5.2 21.6 13.5l68.6 141.3 153.2 22.6c9 1.3 16.5 7.6 19.3 16.3s.5 18.1-5.9 24.5L433.6 328.4l26.2 155.6c1.5 9-2.2 18.1-9.7 23.5s-17.3 6-25.3 1.7l-137-73.2L151 509.1c-8.1 4.3-17.9 3.7-25.3-1.7s-11.2-14.5-9.7-23.5l26.2-155.6L31.1 218.2c-6.5-6.4-8.7-15.9-5.9-24.5s10.3-14.9 19.3-16.3l153.2-22.6L266.3 13.5C270.4 5.2 278.7 0 287.9 0zm0 79L235.4 187.2c-3.5 7.1-10.2 12.1-18.1 13.3L99 217.9 184.9 303c5.5 5.5 8.1 13.3 6.8 21L171.4 443.7l105.2-56.2c7.1-3.8 15.6-3.8 22.6 0l105.2 56.2L384.2 324.1c-1.3-7.7 1.2-15.5 6.8-21l85.9-85.1L358.6 200.5c-7.8-1.2-14.6-6.1-18.1-13.3L287.9 79z",
      ],
      "iconName": "star",
      "prefix": "far",
    }
  }
  inverse={false}
  listItem={false}
  mask={null}
  maskId={null}
  pull={null}
  pulse={false}
  rotation={null}
  shake={false}
  size="2x"
  spin={false}
  spinPulse={false}
  spinReverse={false}
  swapOpacity={false}
  symbol={false}
  title=""
  titleId={null}
  transform={null}
/>
`;

exports[`FaIcon renders icon with \`fa\` wrapper 5`] = `
<FontAwesomeIcon
  beat={false}
  beatFade={false}
  border={false}
  bounce={false}
  className=""
  color="orange"
  fade={false}
  fixedWidth={false}
  flip={false}
  icon={
    Object {
      "icon": Array [
        320,
        512,
        Array [
          10067,
          10068,
          61736,
        ],
        "3f",
        "M80 160c0-35.3 28.7-64 64-64h32c35.3 0 64 28.7 64 64v3.6c0 21.8-11.1 42.1-29.4 53.8l-42.2 27.1c-25.2 16.2-40.4 44.1-40.4 74V320c0 17.7 14.3 32 32 32s32-14.3 32-32v-1.4c0-8.2 4.2-15.8 11-20.2l42.2-27.1c36.6-23.6 58.8-64.1 58.8-107.7V160c0-70.7-57.3-128-128-128H144C73.3 32 16 89.3 16 160c0 17.7 14.3 32 32 32s32-14.3 32-32zm80 320a40 40 0 1 0 0-80 40 40 0 1 0 0 80z",
      ],
      "iconName": "question",
      "prefix": "fas",
    }
  }
  inverse={false}
  listItem={false}
  mask={null}
  maskId={null}
  pull={null}
  pulse={false}
  rotation={null}
  shake={false}
  size="sm"
  spin={false}
  spinPulse={false}
  spinReverse={false}
  swapOpacity={false}
  symbol={false}
  title=""
  titleId={null}
  transform={null}
/>
`;
